### PR TITLE
Add a common autograd TLS state

### PR DESCRIPTION
**Pull request by albanD**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63117
* #63116
* #63115
* __->__ #63114

This PR collapses the GradMode and InferenceMode thread local booleans into a single thread local uint8.
This helps reducing the number of thread local variable accesses done when we propagate ThreadLocalStates.

Note that this is even more beneficial as we will add a forward mode AD TLS (similar to GradMode) higher in this stack and this new structure should reduce the perf impact of adding this new TLS.

Here is the full benchmark result between master and the top of this stack: https://gist.github.com/albanD/e421101e9ed344e94999bef3a54bf0f3
tl;dr: give a benefit in most cases. It is never detrimental.

Differential Revision: [D30388099](https://our.internmc.facebook.com/intern/diff/D30388099)

### CODE REVIEW

**Review by soulitzer — APPROVED** [review]
LGTM! Curious to see what the perf impact is.

**Comment by albanD** [comment]
Updated the main comment with benchmarks.

**Comment by albanD** [comment]
For some reason ghstack didn't like my rebasing work and pushed a brand new stack (maybe I shouldn't have upgraded ghstack?)
In any case, the new version of this PR is here: https://github.com/pytorch/pytorch/pull/63318

Sorry for the noise.

**Review by soulitzer — APPROVED** [review]

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30388099).

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30388099).

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30388099).
